### ISSUE TITLE
[ConvOclDirectFwdGen][FP16] Memory access fault by GPU for certain configurations

### ISSUE THREAD

**Issue by junliume** [ROCmSoftwarePlatform/MIOpen#2522]
This issue appears to be reproducible on all platforms (tested: `gfx1100;gfx908;gfx90a`):

**update**:
the reproduce example is not accurate, it should be:
```
 MIOPEN_FIND_ENFORCE=3 ./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -S -1
```
instead of

```
./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -S 9
```

Observation:
```
Warning: Solution id (9) is not reported by the library. Trying it anyway...
Memory access fault by GPU node-1 (Agent handle: 0x1c27270) on address 0x7f67ace02000. Reason: Page not present or supervisor privilege.
```

- Even by adjusting the configuration a little bit, the solver can actually work, e.g. change H and W to 36

We need to identify the limitation of this solver and update its applicable()

CC: @atamazov and @JehandadKhan

**Comment by JehandadKhan** [comment]
Since the framework is using Find 2.0, can we confirm if the `IsApplicable` is called ? Since in the usual paths ( Immed mode) this solver does not show up.

**Comment by junliume** [comment]
Ping @DrizztDoUrden

**Comment by atamazov** [comment]
@junliume @JehandadKhan The `-S` option in the driver accepts any solverId in Immediate mode, The driver does not call IsApplicable. This is a feature, not a bug, see #638.

Note that driver uses 1.0 API. If we wish to check applicability in 1.0 API, then I would vote for that.

**Comment by DrizztDoUrden** [comment]
Yes, Artem is correct. It doesn't check Applicability with explicitly specified solver id. Find 2.0 uses the same logic. API assumes that solver id or solution received is produced by the library and thus correct. Arbitrary ids are allowed for debug purposes and such. I think we should at least print warning if solver by id is not applicable.

**Comment by DrizztDoUrden** [comment]
The solver may be missing in existing find db record. I would take a look at this specific case. Find 2.0 passes the same tests as find 1.0 so I don't believe that it somehow bypasses isapplicable in find.

**Comment by atamazov** [comment]
@DrizztDoUrden 
> Arbitrary ids are allowed for debug purposes and such

Let me clarify: Arbitrary ids are allowed **_in the driver_** for debug purposes and such.

The library is another story. I would recommend adding IsApplicability check to both 1.0 and 2.0 APIs. We'll discuss this topic with @DrizztDoUrden tomorrow.

**Comment by junliume** [comment]
@DrizztDoUrden @atamazov sorry for the reproduce example not been exactly accurate. One can reproduce the issue by:
```
 MIOPEN_FIND_ENFORCE=3 ./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -S -1
```
Then it will search solver 9 and crash.
The same test will pass if we pass `export MIOPEN_DEBUG_CONV_DIRECT_OCL_FWDGEN=0`

**Comment by atamazov** [comment]
Yes, this is failure of the `MIOpenCDFGen` kernel (solver: `ConvOclDirectFwdGen`).

**Comment by atamazov** [comment]
Reproducible without tuning on gfx1030 by
```
./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -S ConvOclDirectFwdGen
```

**Comment by atamazov** [comment]
Fails on gfx906 with ROCm 5.4.1

**Comment by junliume** [comment]
It's a legacy kernel, we may disable it, but I wish to disable it within the minimal possible cases instead of globally.

**Comment by atamazov** [comment]
Sure. IIRC this kernel is a non-orthodox one and can show very good performance in some useful cases.

**Comment by DrizztDoUrden** [comment]
My proposal is to add IsApplicable checks to both immediate mode and `miopenLoadSolution` for find 2.0. There is not much sense to add it to `miopenRunSolution` as it can only receive it from another API function result and `miopenLoadSolution` is the only one not validating applicability because the binary has been produced by find at some point. I don't think it is of high priority, but rather an enhancement we may want to see at some near enough point in the future. 
But it looks like that is beside the scope of current issue anyway.

**Comment by atamazov** [comment]
@junliume @JehandadKhan 

### Results of testing with different ROCm versions

`develop` b6fe536b9, gfx1030:
- ROCm 5.1, HIP version 5.1.22103 - PASS
- ROCm 5.2, HIP version 5.2.22212 - PASS
- ROCm 5.3, HIP version 5.3.22374 - FAIL
- ROCm 5.3, HIP version 5.3.22374, offline compiler - FAIL (this confirms that COMgr is not involved)
- ROCm 5.5, HIP version 5.7.23365 - FAIL

`develop` 956825adb, gfx906/60
- ROCm 5.4.2, HIP version 5.4.22505 - FAIL

### Some logs
<details><summary><i><b>Log: gfx1030 with ROCm 5.2 - PASS</b></i></summary>

```bash
root@Drakkar:~/github/MLOpen1/build/5.2# MIOPEN_LOG_LEVEL=6 \
> MIOPEN_FIND_MODE=1 \
> MIOPEN_DEBUG_FIND_ONLY_SOLVER=ConvOclDirectFwdGen \
> ./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1030
MIOpen(HIP): Info [Handle] stream: 0xf93c90, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, HIP version 5.2.22212, MIOpen version 3.00.0.
MIOpen(HIP): Info2 [GetWorkSpaceSize]
MIOpen(HIP): Info [GetEnvFindOnlySolverImpl] 9
MIOpen(HIP): Info2 [GetWorkspaceSizes] ConvOclDirectFwdGen: Skipped (no workspace required)
MIOpen(HIP): Info [GetWorkSpaceSize] 0
PRNG seed: 12345678
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 0
MIOpen(HIP): Info [Measure] RamDb::Prefetch time: 0.033213 ms
MIOpen(HIP): Info2 [ValidateUnsafe] DB file is older than cache: 283982893065721, 298731097857815
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F in cache for file /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [Measure] Db::FindRecord time: 0.051286 ms
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF and solver ConvOclDirectFwdGen
MIOpen(HIP): Info2 [LogFindDbItem] Kernel cache entry not found for solver: ConvOclDirectFwdGen at network config: 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F
MIOpen(HIP): Info2 [LogFindDbItem] Find-db record content: ConvOclDirectFwdGen:0.714397,0,miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoWinograd
MIOpen(HIP): Info [GetPerfDbPathFile] Found exact perf database file
MIOpen(HIP): Info2 [SQLiteBase] Initializing system database file /root/github/MLOpen1/src/kernels/gfx1030_36.db
MIOpen(HIP): Info2 [SQLiteBase] Initializing user database file /root/github/MLOpen1/src/kernels/gfx1030_36_1.1.0.udb
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [FindSolutionImpl] ConvOclDirectFwdGen (not searchable)
MIOpen(HIP): Info2 [SearchForAllSolutions] ConvOclDirectFwdGen: Success.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoImplicitGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoFFT
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.2.4.0, USE_HIP_PCH: 1
MIOpen(HIP): Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx1030
MIOpen(HIP): Info2 [Log] Kernel MIOpenConvDirGenFwd.cl Compile Time, ms: 512.653
MIOpen(HIP): Info2 [Log] PrecompileKernels Compile Time, ms: 512.714
MIOpen(HIP): Info2 [PrepareInvoker] Preparing kernel: MIOpenCDFGen
MIOpen(HIP): Info2 [run] kernel_name = MIOpenCDFGen, global_work_dim = { 24, 32, 10 }, local_work_dim = { 8, 16, 1 }
MIOpen(HIP): Info2 [SetValues] 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F, content inserted: ConvOclDirectFwdGen:0.717038,0,miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [EvaluateInvokers] ConvOclDirectFwdGen: MIOpenCDFGen: 0.717038 < 3.40282e+38
MIOpen(HIP): Info2 [Register] Invoker registered for algorithm 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF and solver ConvOclDirectFwdGen
MIOpen(HIP): Info2 [SetAsFound1_0] Solver ConvOclDirectFwdGen registered as find 1.0 best for miopenConvolutionFwdAlgoDirect in 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF
MIOpen(HIP): Info [EvaluateInvokers] Selected: ConvOclDirectFwdGen: MIOpenCDFGen: 0.717038, workspace_sz = 0
MIOpen(HIP): Info2 [StoreRecord] Trying to store record at key 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F in cache for file /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [StoreRecordUnsafe] Storing record: 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F in file /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [FindRecordUnsafe] Key match: 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F
MIOpen(HIP): Info2 [FindRecordUnsafe] Contents found: ConvOclDirectFwdGen:0.714397,0,miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info2 [UpdateDbModificationTime] Updating db modification time for /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [ValidateUnsafe] DB file is newer than cache: 298731612522467, 298731097857815
MIOpen(HIP): Info2 [UpdateDbModificationTime] Updating db modification time for /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [Measure] Db::StoreRecord time: 8.12397 ms
MIOpen(HIP): Info [FindConvolution] miopenConvolutionFwdAlgoDirect      0.717038        0
MIOpen(HIP): Info [FindConvFwdAlgorithm] FW Chosen Algorithm: ConvOclDirectFwdGen , 0, 0.717038
MIOpen(HIP): Info [ConvolutionForward] algo = 1, workspace = 0
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF and algorithm miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info2 [run] kernel_name = MIOpenCDFGen, global_work_dim = { 24, 32, 10 }, local_work_dim = { 8, 16, 1 }
MIOpen Forward Conv. Algorithm: 1, Solution: 9/ConvOclDirectFwdGen
GPU Kernel Time Forward Conv. Elapsed: 0.610519 ms (average)
stats: name, n, c, ho, wo, x, y, k, flopCnt, bytesRead, bytesWritten, GFLOPs, GB/s, timeMs
stats: fwd-conv3x3u2, 1, 260, 17, 17, 3, 3, 308,  416576160, 2078440, 178024, 682, 4, 0.610519
MIOpen(HIP): Info [PrintVersion] HIPRTC v.9.0
Forward Convolution Verifies OK on GPU reference (1.22253e-05 < 0.0082)
```
</details>
<details><summary><i><b>Log: gfx1030 with ROCm 5.3 - FAIL</b></i></summary>

```bash
root@Drakkar:~/github/MLOpen1/build/5.3# MIOPEN_LOG_LEVEL=6 \
> MIOPEN_FIND_MODE=1 \
> MIOPEN_DEBUG_FIND_ONLY_SOLVER=ConvOclDirectFwdGen \
> ./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1030
MIOpen(HIP): Info [Handle] stream: 0xc6aa20, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, HIP version 5.3.22374, MIOpen version 3.00.0.
MIOpen(HIP): Info2 [GetWorkSpaceSize]
MIOpen(HIP): Info [GetEnvFindOnlySolverImpl] 9
MIOpen(HIP): Info2 [GetWorkspaceSizes] ConvOclDirectFwdGen: Skipped (no workspace required)
MIOpen(HIP): Info [GetWorkSpaceSize] 0
PRNG seed: 12345678
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 0
MIOpen(HIP): Info [Measure] RamDb::Prefetch time: 0.026029 ms
MIOpen(HIP): Info2 [ValidateUnsafe] DB file is older than cache: 268938234206983, 268951182368210
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F in cache for file /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [Measure] Db::FindRecord time: 0.048582 ms
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF and solver ConvOclDirectFwdGen
MIOpen(HIP): Info2 [LogFindDbItem] Kernel cache entry not found for solver: ConvOclDirectFwdGen at network config: 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F
MIOpen(HIP): Info2 [LogFindDbItem] Find-db record content: ConvOclDirectFwdGen:0.610298,0,miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoWinograd
MIOpen(HIP): Info [GetPerfDbPathFile] Found exact perf database file
MIOpen(HIP): Info2 [SQLiteBase] Initializing system database file /root/github/MLOpen1/src/kernels/gfx1030_36.db
MIOpen(HIP): Info2 [SQLiteBase] Initializing user database file /root/github/MLOpen1/src/kernels/gfx1030_36_1.1.0.udb
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [FindSolutionImpl] ConvOclDirectFwdGen (not searchable)
MIOpen(HIP): Info2 [SearchForAllSolutions] ConvOclDirectFwdGen: Success.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoImplicitGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoFFT
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.2.4.0, USE_HIP_PCH: 1
MIOpen(HIP): Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx1030
MIOpen(HIP): Info2 [Log] Kernel MIOpenConvDirGenFwd.cl Compile Time, ms: 514.179
MIOpen(HIP): Info2 [Log] PrecompileKernels Compile Time, ms: 514.236
MIOpen(HIP): Info2 [PrepareInvoker] Preparing kernel: MIOpenCDFGen
MIOpen(HIP): Info2 [run] kernel_name = MIOpenCDFGen, global_work_dim = { 24, 32, 10 }, local_work_dim = { 8, 16, 1 }
Memory access fault by GPU node-1 (Agent handle: 0xbe82a0) on address 0x7f9e30a01000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)
```
</details>
<details><summary><i><b>Log: gfx1030 with ROCm 5.7 - FAIL</b></i></summary>

```bash
root@Drakkar:~/github/MLOpen1/build/release.hip-clang# MIOPEN_LOG_LEVEL=6 \
> MIOPEN_FIND_MODE=1 \
> MIOPEN_DEBUG_FIND_ONLY_SOLVER=ConvOclDirectFwdGen \
> ./bin/MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpenDriver convfp16 -n 1 -c 260 -H 35 -W 35 -k 308 -y 3 -x 3 -p 0 -q 0 -u 2 -v 2 -l 1 -j 1 -m conv -g 1 -F 1 -t 1 -i 1
MIOpen(HIP): Info [get_device_name] Raw device name: gfx1030
MIOpen(HIP): Info [Handle] stream: 0x10fe2c0, device_id: 0
MIOpen(HIP): Info [GetFindModeValueImpl] MIOPEN_FIND_MODE = NORMAL(1)
MIOpen(HIP): Info [AmdRocmMetadataVersionDetect] ROCm MD version AMDHSA_COv3, HIP version 5.7.23365, MIOpen version 3.00.0.
MIOpen(HIP): Info2 [GetWorkSpaceSize]
MIOpen(HIP): Info [GetEnvFindOnlySolverImpl] 9
MIOpen(HIP): Info2 [GetWorkspaceSizes] ConvOclDirectFwdGen: Skipped (no workspace required)
MIOpen(HIP): Info [GetWorkSpaceSize] 0
PRNG seed: 12345678
MIOpen(HIP): Info [FindConvFwdAlgorithm] requestAlgoCount = 2, workspace = 0
MIOpen(HIP): Info [Measure] RamDb::Prefetch time: 0.02173 ms
MIOpen(HIP): Info2 [ValidateUnsafe] DB file is older than cache: 266081540719272, 266614448119135
MIOpen(HIP): Info2 [FindRecordUnsafe] Looking for key 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F in cache for file /root/github/MLOpen1/src/kernels/gfx1030_36.HIP.3_00_0.ufdb.txt
MIOpen(HIP): Info2 [Measure] Db::FindRecord time: 0.046695 ms
MIOpen(HIP): Info2 [GetInvoker] Returning an invoker for problem 260x35x35x3x3x308x17x17x1xNCHWxFP16x0x0x2x2x1x1x1xF and solver ConvOclDirectFwdGen
MIOpen(HIP): Info2 [LogFindDbItem] Kernel cache entry not found for solver: ConvOclDirectFwdGen at network config: 260-35-35-3x3-308-17-17-1-0x0-2x2-1x1-0-NCHW-FP16-F
MIOpen(HIP): Info2 [LogFindDbItem] Find-db record content: ConvOclDirectFwdGen:0.610372,0,miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [TryLoad] Find-db regenerating.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoWinograd
MIOpen(HIP): Info [GetPerfDbPathFile] Found exact perf database file
MIOpen(HIP): Info2 [SQLiteBase] Initializing system database file /root/github/MLOpen1/src/kernels/gfx1030_36.db
MIOpen(HIP): Info2 [SQLiteBase] Initializing user database file /root/github/MLOpen1/src/kernels/gfx1030_36_1.1.0.udb
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoDirect
MIOpen(HIP): Info [FindSolutionImpl] ConvOclDirectFwdGen (not searchable)
MIOpen(HIP): Info2 [SearchForAllSolutions] ConvOclDirectFwdGen: Success.
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoImplicitGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoGEMM
MIOpen(HIP): Info2 [Find] Starting find for miopenConvolutionFwdAlgoFFT
MIOpen(HIP): Info [PrintVersionImpl] COMgr v.2.5.0, USE_HIP_PCH: 1
MIOpen(HIP): Info2 [SetIsaName] amdgcn-amd-amdhsa--gfx1030
MIOpen(HIP): Info2 [Log] Kernel MIOpenConvDirGenFwd.cl Compile Time, ms: 512.927
MIOpen(HIP): Info2 [Log] PrecompileKernels Compile Time, ms: 512.964
MIOpen(HIP): Info2 [PrepareInvoker] Preparing kernel: MIOpenCDFGen
MIOpen(HIP): Info2 [run] kernel_name = MIOpenCDFGen, global_work_dim = { 24, 32, 10 }, local_work_dim = { 8, 16, 1 }
Memory access fault by GPU node-1 (Agent handle: 0x179b160) on address 0x7fb861000000. Reason: Page not present or supervisor privilege.
Aborted (core dumped)

```
</details>

### Conclusion: the reason of the issue is some regression in ROCm 5.3. 
Likely root causes (in descending order of likelihood):
- (1) OpenCL LLVM compiler regression
- ~(2) Problem in COMgr~
- (3) HIP runtime regression with binaries produced from OpenCL sources.
- (4) Ill-formed OpenCL kernel (unspecified or undefined behavior).

So far this should be qualified as https://github.com/ROCmSoftwarePlatform/MIOpen/labels/non-miopen-bug

**Comment by atamazov** [comment]
### Proposed W/A

Starting from ROCm 5.3, disable the faulty solver for FP16 configs that have stride 2x2 ~!= 1x1~.

**Comment by atamazov** [comment]
The same problems happen with BF16 datatype.

**Comment by atamazov** [comment]
@JehandadKhan I've investigated the source code in src/kernels/MIOpenConvDirGenFwd.cl. Mempory access failure happens at line 168:
```C++
lcl_wei[i] =
                weights[wei_off + lcl_o * MLO_N_IN_CHNLS * MLO_FLTR_SZ0 * MLO_FLTR_SZ1 + lcl_o_i];
```
**_It is interesting that the same FP32 configs do not fail._** Taking into account that neither the solver nor the kernel use the _size of the datatype_ for computations of addresses/indices etc, I would suspect that compiler produces somewhat incorrect code for half-DWORD (16-bit) memory reads.

Also I've played with some suspicious kernel code (which, of course, _is there_), but without success.

**Comment by atamazov** [comment]
@junliume @JehandadKhan The root reason of the [out-of-bounds memory access](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2522#issuecomment-1821565260) is now identified. This is a bug in the kernel, i.e. the root reason falls into category defined in bullet (4) [in this comment](https://github.com/ROCmSoftwarePlatform/MIOpen/issues/2522#issuecomment-1813352890).

The issue should be re-qualified as https://github.com/ROCmSoftwarePlatform/MIOpen/labels/bug

The fix in #2546 is ready for review, testing and merging.

**Comment by atamazov** [comment]
@junliume @JehandadKhan #2546 merged, let's close this.
